### PR TITLE
chore: pipeline component refactor

### LINKED ISSUE #777: Pipeline refactor

### PR DESCRIPTION
**Pull request by veds-g**
Fixes #777 
- Separated data fetching, formatting and rendering logic
- Removed redundant API calls made to fetch data for rendering pipeline
- Improved the number of re-renders made due to multiple API calls
- Fixed the intended refresh rate for API calls, currently refreshing every 30sec but the intention is 30sec / 60sec for different APIs
- Added a loading state for graph rendering
image

### CODE REVIEW

**Comment by vigith** [comment]
> * Refresh pipeline information every 30 secs

are we seeing issues with the current 15 sec refresh?

**Comment by vigith** [comment]
> > are we seeing issues with the current 15 sec refresh?
> 
> Currently, the attempt in UI code is a 30-sec refresh for pipeline info and a 60-sec refresh for metrics and watermark. Due to an incorrect approach, everything refreshes after 30-secs now. To fix the approach intended as per the comments in UI code I modified the refresh windows.

Ok, it is a bug fix. Could you please update the description?

**Comment by KeranYang** [comment]
Hey @veds-g, I am trying to learn about the UI part of Numaflow. Could you comment on the code changes that reflect each of the bullet points in PR description? Looking at the Files changed section I can only tell it fixes the first point. Thanks!

**Comment by veds-g** [comment]
> Hey @veds-g, I am trying to learn about the UI part of Numaflow. Could you comment on the code changes that reflect each of the bullet points in PR description? Looking at the Files changed section I can only tell it fixes the first point. Thanks!

Hey @KeranYang, regarding the points mentioned in the description - the last point is actually covered in the comments in `utils/fetchWrappers/pipelineViewFetch.ts` (here the refresh intervals are mentioned above useEffect hooks). Regarding the other points, they are more of an approach changes taken to reduce re-renders based on the **state-management** which is a part of how React components should be defined, so I can help you understand the flow separately which should give you a good idea about how these issues are solved!

File: ui/src/components/pages/Pipeline/index.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  }, [watermarkErr]);
+
+  if (pipelineErr || buffersErr) {
+    return <div>Error</div>;
+  }
+
+  return (
+    <div data-testid={"pipeline"} style={{ overflow: "scroll !important" }}>
+      {pipeline?.spec &&
+        buffers.length > 0 &&
```
Comment (mshakira) [inline ui/src/components/pages/Pipeline/index.tsx:56]: please do null checks

File: ui/src/components/pages/Pipeline/index.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import "./style.css";
+
+export function Pipeline() {
+  const { namespaceId, pipelineId } = useParams();
+  const {
+    pipeline,
+    buffers,
+    vertices,
+    edges,
```
Comment (mshakira) [inline ui/src/components/pages/Pipeline/index.tsx:16]: do we need loading state? what will happen when no data is returned

File: ui/src/utils/fetchWrappers/pipelineViewFetch.ts [hunk 3]
```diff
@@ -0,0 +1,400 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
```
Comment (mshakira) [inline ui/src/utils/fetchWrappers/pipelineViewFetch.ts:1]: Can we rename the dir with hooks?

Comment (mshakira) [reply ui/src/utils/fetchWrappers/pipelineViewFetch.ts:1]: fetcherHooks may be

Comment (veds-g) [reply ui/src/utils/fetchWrappers/pipelineViewFetch.ts]: redefined fetcherHooks under Utils separately

**Review by mshakira — APPROVED** [review]

**Comment by vigith** [comment]
@veds-g is this PR ready?

**Review by yhl25 — DISMISSED** [review]
Screenshot 2023-06-21 at 9 43 12 PM

not able to load the pods view, getting 500. Can you please fix it?

pipeline spec
```apiVersion: numaflow.numaproj.io/v1alpha1
kind: Pipeline
metadata:
  name: simple-pipeline
spec:
  vertices:
    - name: in
      scale:
        min: 1
      source:
        generator:
          keyCount: 100
          msgSize: 100
          rpu: 10
          duration: 1s
    - name: first-cat
      scale:
        min: 2
      udf:
        builtin:
          name: cat
    - name: second-cat
      scale:
        min: 2
      udf:
        builtin:
          name: cat
    - name: counter
      partitions: 2
      udf:
        container:
          image: quay.io/yhl25/counter:v1
        groupBy:
          window:
            fixed:
              length: 60s
          keyed: true
          storage:
            persistentVolumeClaim:
              volumeSize: 2Gi
              accessMode: ReadWriteOnce
    - name: out
      scale:
        min: 2
      sink:
        log: {}
  edges:
    - from: in
      to: first-cat
    - from: first-cat
      to: second-cat
    - from: second-cat
      to: counter
    - from: counter
      to: out```

**Comment by veds-g** [comment]
@yhl25 @vigith 
I tried this pipeline in k3d cluster and iks cluster, works just fine. Only see this error in kind cluster.

**Pull request merged**